edge-util/test/Microsoft.Azure.Devices.Edge.Storage.RocksDb.Test/ColumnFamilyStoreTest.cs

```diff
@@ -103,5 +104,53 @@ public async Task MessageCountTest()
                 Assert.Equal(0ul, await columnFamilyDbStore.Count());
             }
         }
+
+        [Fact]
+        public async Task MessageCountUnderflowTest()
+        {
+            using (IDbStore columnFamilyDbStore = this.rocksDbStoreProvider.GetDbStore("test"))
+            {
+                Assert.Equal(0ul, await columnFamilyDbStore.Count());
+
+                for (int i = 0; i < 10; i++)
+                {
+                    string key = $"key{i}";
+                    string value = "$value{i}";
+                    await columnFamilyDbStore.Put(key.ToBytes(), value.ToBytes());
+                }
+
+                Assert.Equal(10ul, await columnFamilyDbStore.Count());
+            }
+
+            using (IDbStore columnFamilyDbStore = this.rocksDbStoreProvider.GetDbStore("test"))
+            {
+                Assert.Equal(10ul, await columnFamilyDbStore.Count());
+
+                // Using 11 (10 + 1) to make sure underflow is caught.
+                for (int i = 0; i < 11; i++)
+                {
+                    string key = $"key{i}";
+                    await columnFamilyDbStore.Remove(key.ToBytes());
+                }
+
+                Assert.Equal(0ul, await columnFamilyDbStore.Count());
+            }
+
+            // Verify with threads.
+
+            using (IDbStore columnFamilyDbStore = this.rocksDbStoreProvider.GetDbStore("test"))
+            {
+                for (int i = 0; i < 100; i++)
+                {
+                    var t = new Thread(async () =>
+                    {
+
+                        string key = $"key{0}";
+                        await columnFamilyDbStore.Remove(key.ToBytes());
+                        Assert.Equal(0ul, await columnFamilyDbStore.Count());
+                    });
+                }
+            }
+        }
```

Comment:
Not sure if this is the best way to try and test this... As with any race condition this is kinda 'luck' based, which don't make good tests. If there is a good utility in C# or pattern that can be used, please let me know!